### PR TITLE
Model config model migration

### PR DESCRIPTION
**Pull request by SimonRichardson**
Adds model config model migration to the domain model config worker flow. This is required to ensure that model migrations succeed. Provider trackers require that it has a valid model config when attempting to start up.

The code is relatively simple by just using import and exporting methods available to us we can set and retrieve config.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd src
$ juju add-model default
$ juju deploy ubuntu

$ juju bootstrap lxd dst
$ juju switch src
$ juju migrate src:default dst

$ juju switch dst:default
$ juju add-machine
```

## Links

**Jira card:** JUJU-5836

### CODE REVIEW

**Comment by nvinuesa** [comment]
QA passes when migrating 4.0->4.0, but when trying 3.5->4.0 I get:
```
machine-0: 12:00:42 ERROR juju.worker.dependency "log-forwarder" manifold worker returned unexpected error: model "7497e627-d853-429a-819d-9c6098032e8a": settings not found (not found)
machine-0: 12:00:42 ERROR juju.worker.dependency "environ-tracker" manifold worker returned unexpected error: model "m" (7497e627-d853-429a-819d-9c6098032e8a): reading model config: model "7497e627-d853-429a-819d-9c6098032e8a": settings not found (not found)
machine-0: 12:00:42 INFO juju.worker.migrationmaster.7497e6 logger-tags:migration setting migration phase to DONE
machine-0: 12:00:42 ERROR juju.worker.dependency "migration-master" manifold worker returned unexpected error: failed to set phase: could not get migration: model "7497e627-d853-429a-819d-9c6098032e8a" not found (not found)
machine-0: 12:00:42 ERROR juju.rpc error writing response: *websocket.netError write tcp 127.0.0.1:17070->127.0.0.1:35428: write: connection reset by peer
machine-0: 12:00:42 ERROR juju.rpc error writing response: *websocket.netError write tcp 127.0.0.1:17070->127.0.0.1:35428: write: connection reset by peer
machine-0: 12:00:42 ERROR juju.rpc error writing response: *websocket.netError write tcp 127.0.0.1:17070->127.0.0.1:35428: write: connection reset by peer
machine-0: 12:00:42 ERROR juju.rpc error writing response: *websocket.netError write tcp 127.0.0.1:17070->127.0.0.1:35428: write: connection reset by peer
machine-0: 12:00:42 ERROR juju.apiserver error serving RPCs: codec.ReadHeader error: error receiving message: write tcp 127.0.0.1:17070->127.0.0.1:35428: write: connection reset by peer
```
on the source controller (the dest controller rebounces).

**Comment by SimonRichardson** [comment]
@nvinuesa We'll still need more of @tlm work before it will work for 3.5 -> 4.0. I'm happy with 4.0 -> 4.0 though.

**Review by nvinuesa — APPROVED** [review]
Code LGTM and QA passes (on 4.0->4.0 only for the moment)

File: core/modelmigration/pacage_test.go [hunk 1]
```diff
@@ -30,6 +30,11 @@ func (*ImportTest) TestImports(c *gc.C) {
 	// This package should only depend on other core packages.
 	// If this test fails with a non-core package, please check the dependencies.
 	c.Assert(found, jc.SameContents, []string{
```
Comment (manadart) [inline core/modelmigration/pacage_test.go:32]: Can you fix the name of this file? It's _pacage_test.go_.

File: core/modelmigration/pacage_test.go [hunk 2]
```diff
@@ -30,6 +30,11 @@ func (*ImportTest) TestImports(c *gc.C) {
...
+		"core/credential",
 		"core/database",
+		"core/life",
+		"core/model",
+		"core/user",
+		"internal/uuid",
```
Comment (manadart) [inline core/modelmigration/pacage_test.go:38]: Technically shouldn't allow this.

File: domain/modelconfig/modelmigration/export.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+// exportOperation describes a way to execute a migration for
+// exporting external controllers.
+type exportOperation struct {
+	modelmigration.BaseOperation
+
+	service ExportService
+}
+
+func (e *exportOperation) Setup(scope modelmigration.Scope) error {
```
Comment (manadart) [inline domain/modelconfig/modelmigration/export.go:39]: Comment.

File: domain/modelconfig/modelmigration/import.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// RegisterImport registers the import operations with the given coordinator.
+func RegisterImport(coordinator Coordinator, defaultsProvider service.ModelDefaultsProvider) {
+	coordinator.Add(&importOperation{
+		defaultsProvider: defaultsProvider,
+	})
+}
+
+type ImportService interface {
```
Comment (manadart) [inline domain/modelconfig/modelmigration/import.go:31]: And here.

File: domain/modelconfig/modelmigration/import.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+type importOperation struct {
+	modelmigration.BaseOperation
+
+	service          ImportService
+	defaultsProvider service.ModelDefaultsProvider
+}
+
+func (i *importOperation) Setup(scope modelmigration.Scope) error {
```
Comment (manadart) [inline domain/modelconfig/modelmigration/import.go:48]: Etc.

File: internal/database/errors.go [hunk 6]
```diff
@@ -19,45 +19,104 @@ import (
 // IsErrConstraintCheck returns true if the input error was
 // returned by SQLite due to violation of a constraint check.
 func IsErrConstraintCheck(err error) bool {
+	if err == nil {
```
Comment (manadart) [inline internal/database/errors.go:22]: Couldn't these all be a one liner with `||`?

**Review by manadart — APPROVED** [review]
But for the nits, this is OK. I'm relying on @nvinuesa's confirmation of QA.

File: core/modelmigration/pacage_test.go (hunk 2 above)
Comment (SimonRichardson) [reply core/modelmigration/pacage_test.go:38]: If we outsourced to `google/uuid` then this goes away.

**Pull request merged**